### DISCUSSION TITLE
[Help Wanted] cannot follow Hints for `<button>` injected by userscript

### DISCUSSION THREAD

**Discussion by mokurin000** [qutebrowser/qutebrowser#6408] [Q&A]
**Version info**:

qutebrowser v2.1.0
Git commit: 368ed9a4b on v2.1.x (2021-03-12 13:57:57 +0100)
Backend: QtWebEngine 5.15.2, Chromium 83.0.4103.122
Qt: 5.15.2

**Does the bug happen if you start with `--temp-basedir`?**:
never

**Description**

the \<button\> injected by userscript cannot click though `f` command

**How to reproduce**

https://huodong.xueanquan.com/2021gdgjaq/video.html (This page is in Chinese)
xuesheng : 123456 is fine
userscript is following.
[anquan.js.gz](https://github.com/qutebrowser/qutebrowser/files/6331333/anquan.js.gz)

```javascript
// ==UserScript==
// @name         安全教育平台自动答题
// @description  支持安全课程、专题活动
// @namespace    https://greasyfork.org/users/197529
// @version      0.8.3
// @author       kkocdko
// @license      Unlicense
// @match        *://*.xueanquan.com/*
// @noframes
// ==/UserScript==
"use strict";

const { addFloatButton } = {
  addFloatButton(text, onClick) /* 20200707-123713 */ {
    if (!document.addFloatButton) {
      const container = document.body
        .appendChild(document.createElement("div"))
        .attachShadow({ mode: "open" });
      container.innerHTML =
        "<style>:host{position:fixed;top:3px;left:3px;z-index:2147483647;height:0}#i{display:none}*{float:left;margin:4px;padding:1em;outline:0;border:0;border-radius:5px;background:#1e88e5;box-shadow:0 1px 4px rgba(0,0,0,.1);color:#fff;font-size:14px;line-height:0;transition:.3s}:active{background:#42a5f5;box-shadow:0 2px 5px rgba(0,0,0,.2)}button:active{transition:0s}:checked~button{visibility:hidden;opacity:0;transform:translateY(-3em)}label{border-radius:50%}:checked~label{opacity:.3;transform:translateY(3em)}</style><input id=i type=checkbox><label for=i></label>";
      document.addFloatButton = (text, onClick) => {
        const button = document.createElement("button");
        button.textContent = text;
        button.addEventListener("click", onClick);
        return container.appendChild(button);
      };
    }
    return document.addFloatButton(text, onClick);
  },
};

addFloatButton("Auto Answer", () => {
  document.querySelectorAll(":enabled:not(:checked)").forEach((el) => {
    el.value = 1;
    el.click();
  });
});

addFloatButton("Semi-Auto Answer", () => {
  document.querySelectorAll("[type=radio],[type=checkbox]").forEach((el) => {
    if (el.checked) return;
    el.value = 1;
    el.click();
  });
  scroll({ top: 9e9 });
});
```

**Comment by The-Compiler** [comment]
Huh, I commented on this but now my comment is gone :thinking: 

I can't see any new button when I visit that page with the userscript.

Can you try to remove the `.attachShadow({ mode: "open" })` and see if it works then?

**Comment by mokurin000** [comment]
> 
> 
> Huh, I commented on this but now my comment is gone 🤔
> 
> I can't see any new button when I visit that page with the userscript.
> 
> Can you try to remove the `.attachShadow({ mode: "open" })` and see if it works then?

That userscript is a piece of this one...

**Comment by The-Compiler** [comment]
I downloaded the full userscript via the `.gz` file, but I still don't see a difference on the page with it installed.

**Comment by mokurin000** [comment]
> 
> 
> I downloaded the full userscript via the `.gz` file, but I still don't see a difference on the page with it installed.

let me try
![qutebrowser_oodNCVS38z](https://user-images.githubusercontent.com/34085039/115220050-b31b0680-a13a-11eb-9d23-115b52c99f91.png)
Could you see these buttons?

**Comment by The-Compiler** [comment]
Ah, thanks for the screenshot - I was looking for something gray (like an unstyled button) and missed those.

**Comment by The-Compiler** [comment]
So the problem here is the `.attachShadow({ mode: "open" })` which creates a [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM), and qutebrowser can't access the elements in there.

Nothing qutebrowser can do about this unfortunately - and removing the `.attachShadow({ mode: "open" })` unfortunately seems to break the buttons. I'll convert this to a discussion, maybe someone has an idea how to change the script so it works.

**Reply by The-Compiler** [reply]
After taking a closer look, I think qutebrowser could actually access shadow DOMs with the "open" mode: https://github.com/qutebrowser/qutebrowser/issues/3569#issuecomment-822374249

**Reply by The-Compiler** [reply]
No, the issues really is the shadow DOM. With it, qutebrowser currently doesn't see the elements at all.
